distributed/scheduler.py

```diff
@@ -1824,24 +1830,24 @@ def decide_worker_rootish_queuing_disabled(
         if not pool:
             return None
 
-        lws = self.last_root_worker
+        tg = ts.group
+        lws = tg.last_worker
         if not (
-            lws
-            and self.last_root_worker_tasks_left
-            and self.workers.get(lws.address) is lws
+            lws and tg.last_worker_tasks_left and self.workers.get(lws.address) is lws
         ):
             # Last-used worker is full or unknown; pick a new worker for the next few tasks
-            ws = self.last_root_worker = min(
-                pool, key=lambda ws: len(ws.processing) / ws.nthreads
-            )
-            # TODO better batching metric (`len(tg)` is not necessarily the total number of root tasks!)
-            self.last_root_worker_tasks_left = math.floor(
-                (len(ts.group) / self.total_nthreads) * ws.nthreads
+            ws = min(pool, key=partial(self.worker_objective, ts))
+            tg.last_worker_tasks_left = math.floor(
+                (len(tg) / self.total_nthreads) * ws.nthreads
             )
         else:
             ws = lws
 
-        self.last_root_worker_tasks_left -= 1
+        # Record `last_worker`, or clear it on the final task
+        tg.last_worker = (
+            ws if tg.states["released"] + tg.states["waiting"] > 1 else None
+        )
+        tg.last_worker_tasks_left -= 1
```

Comment:
For reference, what it looked like before #6985: https://github.com/dask/distributed/blob/f07f384eb0b8b6bad7518e8535e42fdba106ed4b/distributed/scheduler.py#L1850-L1868

Note there's still a slight change here—previously we'd schedule onto any worker if none were idle, now we only schedule onto running workers.